### PR TITLE
Support bounds widening for conditionals within switch statements

### PR DESCRIPTION
**Pull request by mgrang**
Added support for widening bounds for nt_array_ptr dereferences in switch
statements. For example, "switch (*p) {case 'a': break;}" would widen the
bounds of p inside the "case 'a'" block.

### CODE REVIEW

**Comment by mgrang** [comment]
This follows from the PR on for loops: #804

**Comment by mgrang** [comment]
> I have suggestions for improving and generalizing the handling of case statements.
> 
> Could you open a GitHub issue to allow widening for the default case in some circumstances? That should be a separate change. If we establish that another label in a switch statement tests for 0, then the default case will handle non-zero case, and the bounds can be widened there. Programmers will use all the different forms of case statements and may depend on that behavior.

Opened issue: https://github.com/microsoft/checkedc-clang/issues/818

File: clang/lib/Sema/BoundsAnalysis.cpp [hunk 1]
```diff
@@ -71,9 +75,21 @@ void BoundsAnalysis::ComputeGenSets() {
...
   // "if (*(p + i))" then Gen[B1] = {B2, p:i} .
 
+  // Here, EB is B2.
   for (const auto item : BlockMap) {
     ElevatedCFGBlock *EB = item.second;
 
+    // Check if this is a switch case and whether it tests for a null
+    // terminator, like "case '\0'". We cannot widen the bounds for the null
+    // terminator case.
+    bool IsSwitchCaseNullTerm = false;
```
[the remark this answers is not in the document]
Comment (mgrang) [reply clang/lib/Sema/BoundsAnalysis.cpp]: My updated PR generalizes the checking of non-null case labels. Also added more unit tests for switch-cases.

File: clang/lib/Sema/BoundsAnalysis.cpp [hunk 2]
```diff
@@ -83,13 +87,71 @@ void BoundsAnalysis::InitInOutSets() {
...
+
+  // case '1': This is how it is represented in the CFG:
+  // ConstantExpr 'int'
+  //   -CharacterLiteral 'int' 49
+  // This is an IntegerConstantExpr. After the call to isIntegerConstantExpr,
+  // the variable IntVal would contain the value '1'.
+
+  llvm::APSInt IntVal (Ctx.getTypeSize(Ctx.IntTy), 0);
+
+  if (const auto *CE = dyn_cast_or_null<ConstantExpr>(E)) {
```
Comment (dtarditi) [inline clang/lib/Sema/BoundsAnalysis.cpp:111]: Why are you casting E to a ConstantExpr?   Can't you just use the inner test?

File: clang/lib/Sema/BoundsAnalysis.cpp [hunk 3]
```diff
@@ -83,13 +87,71 @@ void BoundsAnalysis::InitInOutSets() {
...
+    if (!E->isIntegerConstantExpr(IntVal, Ctx))
+      IntVal = CE->getResultAsAPSInt();
+  }
+  return IntVal;
+}
+
+bool BoundsAnalysis::CheckIsSwitchCaseNull(ElevatedCFGBlock *EB) {
+  if (const auto *CS = dyn_cast_or_null<CaseStmt>(EB->Block->getLabel())) {
+    llvm::APSInt Zero (Ctx.getTypeSize(Ctx.IntTy), 0);
```
Comment (dtarditi) [inline clang/lib/Sema/BoundsAnalysis.cpp:120]: I believe this may not be the correct type for zero.   The C11 specification says in 6.8.4.2 that the type of the constant expression in a case expression shall be an integer constant expression.   There are a number of different sizes of integer types in C, which are distinct from the `int` type.  The C11 specification says that integer promotions are applied to the controlling expression in the switch statement, and that the label expressions of the case statements are promoted to match that type.   

I believe that the comparison functions for APSInt  may expect the integers to have the same size.   It's been a while since I used the API, so I amy be wrong. In that case, the test for inclusion of zero, as well as the uses of APSInt in GetSwitchCaseVal could both lead to internal compiler assertion failures, if the integer requires more bits than IntTy.

My suggestion is that you test out switches that use an `unsigned int`, as well as an integer type larger than the default size of `int`.  I believe the default promotions rules for `char` will widen everything out to `int`, which is why the code is working for your test cases.

File: clang/lib/Sema/BoundsAnalysis.cpp [hunk 4]
```diff
@@ -103,10 +165,33 @@ void BoundsAnalysis::ComputeGenSets() {
       // Here we have the edges (B1->B2) and (B1->B3). We can add "p:i" only
       // on the true edge. Which means we will add the following entry to
       // Gen[B1]: {B2, p:i}
-      if (!pred->succ_empty() && *pred->succs().begin() == EB->Block)
+
+      // Check if EB is on a true edge of pred. The false edge (including the
+      // default case for a switch) is always the last edge in the list of
+      // edges. So we check whether EB is on the last edge for pred.
```
Comment (dtarditi) [inline clang/lib/Sema/BoundsAnalysis.cpp:171]: The wording of the comment "We check whether EB is on the last edge for pred" is confusing" with respect to the code.  Aren't you checking that EB is not the last edge for pred?

File: clang/lib/Sema/BoundsAnalysis.cpp [hunk 5]
```diff
@@ -103,10 +165,33 @@ void BoundsAnalysis::ComputeGenSets() {
...
+      if (pred->succ_size() &&
+          EB->Block != *(pred->succs().end() - 1)) {
         // Get the edge condition and fill the Gen set.
-        if (Expr *E = GetTerminatorCondition(pred))
+        if (Expr *E = GetTerminatorCondition(pred)) {
+
+          // Check if the pred ends in a switch statement.
```
Comment (dtarditi) [inline clang/lib/Sema/BoundsAnalysis.cpp:177]: Could you add a TODO about handling the default case better and cite the GitHub issue?

File: clang/lib/Sema/BoundsAnalysis.cpp [hunk 6]
```diff
@@ -83,13 +87,71 @@ void BoundsAnalysis::InitInOutSets() {
...
+    return IntVal;
+
+  // const int i = 1;
+  // case i: This is how it is represented in the CFG:
+  // ConstantExpr 'const int' lvalue 1
+  //   -DeclRefExpr 'const int' lvalue Var 'i' 'const int'
+  // This is not an IntegerConstantExpr. It is a ConstantExpr and we get its
+  // value by calling the getResultAsAPSInt() method.
+  if (const auto *CE = dyn_cast_or_null<ConstantExpr>(E))
+    return CE->getResultAsAPSInt();
```
Comment (mgrang) [inline clang/lib/Sema/BoundsAnalysis.cpp:113]: getResultAsAPSInt can only be called on a ConstantExpr. So I have to cast E to ConstantExpr here.

File: clang/lib/Sema/BoundsAnalysis.cpp (hunk 3 above)
Comment (mgrang) [reply clang/lib/Sema/BoundsAnalysis.cpp:120]: I have added new tests for unsigned, UINT_MAX, INT_MAX and INT_MIN case labels. The compiler and GetSwitchCaseVal seem to handle these correctly. For example:

`case 999999999999999999999999999` is internally represented as `11515845246265065471ULL` and we return detect that this is non-zero.

In case of `case INT_MIN + INT_MIN` we do not widen as the compiler evaluates this to 0 with the warning: "overflow in expression; result is 0 with type 'int'".

File: clang/lib/Sema/BoundsAnalysis.cpp (hunk 5 above)
Comment (mgrang) [reply clang/lib/Sema/BoundsAnalysis.cpp:183]: Done.

File: clang/lib/Sema/BoundsAnalysis.cpp [hunk 7]
```diff
@@ -83,13 +87,76 @@ void BoundsAnalysis::InitInOutSets() {
...
+  // case i: This is how it is represented in the CFG:
+  // ConstantExpr 'const int' lvalue 1
+  //   -DeclRefExpr 'const int' lvalue Var 'i' 'const int'
+  // This is not an IntegerConstantExpr. It is a ConstantExpr and we get its
+  // value by calling the getResultAsAPSInt() method.
+
+  // Note: According to C11 spec sections 6.6 and 6.8.1 a DeclRefExpr is not
+  // considered a constant expr. But clang allows for this additional
+  // extension. So we handle this here.
+  if (const auto *CE = dyn_cast_or_null<ConstantExpr>(E))
```
[the remark this answers is not in the document]
Comment (mgrang) [reply clang/lib/Sema/BoundsAnalysis.cpp:116]: Consider this case:
`const int i = 1;`

This is represented in AST as:
`ConstantExpr 'const int' lvalue 1`
`  -- DeclRefExpr 'const int' lvalue Var 'i' 'const int'`

The case for ConstantExpr invokes CheckICE on the DeclRefExpr. Now DeclRefExprs only seem to be considered ICE for EnumConstantDecl or when the language is C++.

```

That's the reason I had a separate case for ConstantExpr.
case Expr::DeclRefExprClass: {
     if (isa<EnumConstantDecl>(cast<DeclRefExpr>(E)->getDecl()))
       return NoDiag();
     const ValueDecl *D = cast<DeclRefExpr>(E)->getDecl();
     if (Ctx.getLangOpts().CPlusPlus &&
         D && IsConstNonVolatile(D->getType())) {
     ...
     }
     return ICEDiag(IK_NotICE, E->getBeginLoc());
```

[the remark this answers is not in the document]
Comment (dtarditi) [reply clang/lib/Sema/BoundsAnalysis.cpp:116]: Thank you for clarifying this and giving this concrete example.  Yes, it makes sense to keep this case.   In test cases where this non-standard extension is used, could you call out that the usage is non-standard C, but allowed by clang.

[the remark this answers is not in the document]
Comment (mgrang) [reply clang/lib/Sema/BoundsAnalysis.cpp:116]: Thanks David. 
And as a result of keeping this case we also need to strip off casts because `getResultAsAPSInt` does not recursively test subexprs.

```
llvm::APSInt ConstantExpr::getResultAsAPSInt() const {
  switch (ConstantExprBits.ResultKind) {
  case ConstantExpr::RSK_APValue:
    return APValueResult().getInt();
  case ConstantExpr::RSK_Int64:
    return llvm::APSInt(llvm::APInt(ConstantExprBits.BitWidth, Int64Result()),
                        ConstantExprBits.IsUnsigned);
  default:
    llvm_unreachable("invalid Accessor");
  }
}
```

I can move the stripping off casts to just above the call to `getResultAsAPSInt`.

[the remark this answers is not in the document]
Comment (dtarditi) [reply clang/lib/Sema/BoundsAnalysis.cpp:116]: It makes sense to move stripping off casts earlier.   I am still concerned that stripping off casts could cause our analysis to differ from what  being tested at runtime.   What does the code generation phase do for this case?  How are they handling casts?

[the remark this answers is not in the document]
Comment (dtarditi) [reply clang/lib/Sema/BoundsAnalysis.cpp:116]: I meant "move stripping off casts later".

[the remark this answers is not in the document]
Comment (mgrang) [reply clang/lib/Sema/BoundsAnalysis.cpp:116]: The overflown value is truncated to the valid integer range in clang ("lib/Sema/SemaStmt.cpp"):
```
    llvm::APSInt ConvVal(Val);
    AdjustAPSInt(ConvVal, UnpromotedWidth, UnpromotedSign);
    AdjustAPSInt(ConvVal, Val.getBitWidth(), Val.isSigned());
    if (ConvVal != Val)
      S.Diag(Loc, diag::warn_case_value_overflow);

static void AdjustAPSInt(llvm::APSInt &Val, unsigned BitWidth, bool IsSigned) {
  Val = Val.extOrTrunc(BitWidth);
  Val.setIsSigned(IsSigned);
}
```

So in the AST there is no cast for the trunc/ext as the AST does not record the trunc/ext.
And the codegen simply receives the trunc/ext value. Here's the codegen for the following reduced test case:
```
void f(unsigned int a) {
  switch (a) {
  case 0x500000000LL:
    break;
  }
}
```
Codegen:
```
        sub     sp, sp, #4
        str     r0, [sp]
        ldr     r0, [sp]
        cmp     r0, #0
        bne     .LBB0_2
        b       .LBB0_1
.LBB0_1:                                @ %sw.bb
        b       .LBB0_2
.LBB0_2:                                @ %sw.epilog
        add     sp, sp, #4
```

As we can see the value of the label (cmp r0, #0) is 0.
